Encryption.md

```diff
@@ -19,216 +19,544 @@
+is set to 1. 
+
+### 4.3 Key metadata
+A wide variety of services and tools for management of encryption keys exist in the 
+industry today. Public clouds offer different key management services (KMS), and 
+organizational IT systems either build proprietary key managers in-house or adopt open source 
+tools for on-premises deployment. Besides the diversity of management tools, there are many 
+ways to generate and handle the keys themselves (generate Data keys inside KMS – or locally 
+upon data encryption; use Data keys only, or use Master keys to encrypt/wrap the Data keys; 
+store the wrapped key material inside the data file, or at a separate location; etc). There 
+is also a large variety of authorization and certification methods, required to control the 
+access to encryption keys.
+
+Parquet is not limited to a single KMS, key generation/wrapping method, or authorization service. 
+Instead, Parquet provides a developer with a simple interface that can be utilized for implementation 
+of any key management scheme. For each column or footer key, a file writing code can pass an 
+arbitrary `key_metadata` byte array that will be stored in the file. A file reading code can pass a 
+key retrieval callback, that will invoked on the key_metadata byte array in order to recover the
+encryption (or signature) key. For example, the key_metadata can keep a serialized
+
+   * String ID of a Data encryption key (for direct key retrieval from a KMS).
+   * Wrapped Data key, and string ID of a Master key. The Data key is generated randomly and 
+   wrapped either remotely in a KMS, or locally after retrieving the Master key from KMS. Wrapping 
+   format depends on the KMS and can be a JSON string, or a base64 encoded byte array in a case of 
+   local wrapping.
+   * Short ID (counter) of a key inside the Parquet data file. The Data key is wrapped with a 
+   Master key using one of the options described above – but the resulting key material is stored 
+   separately, outside the data file, and will be retrieved using the counter and file path.
+   * Any of the three above, plus a string ID of the KMS instance (in case there are many).
+   * Random string - useful for creation of footer keys in organizations not willing to authenticate 
+   and connect all readers to a KMS. The footer key is generated as F(key_metadata), where F is an 
+   algorithm unknown outside the organization. These keys are less secure than KMS-managed keys, but 
+   better than using no keys at all in plaintext and unsigned footers (discussed in the sections 5.2 
+   and 5.3.)
+   
+Key metadata can also be empty - in a case the encryption keys are fully managed by the caller 
+code, and passed explicitly to Parquet for the footer and each encrypted column.
+
+### 4.4 Additional Authenticated Data
+The AES GCM cipher in the basic mode protects against byte replacement inside a ciphertext, 
```

Comment:
This should not use the term "basic mode". That was the "default" mode earlier, but both terms are not clear. Instead, if you need to refer to GCM without an AAD, just say "without passing AAD". That is specific and will not be misunderstood.